File: code/ch01/1.3.6.permute.js
function backtrack(list, tempList, nums) {
  if (tempList.length === nums.length) return list.push([...tempList]);
  for (let i = 0; i < nums.length; i++) {
    if (tempList.includes(nums[i])) continue;
    tempList.push(nums[i]);
    backtrack(list, tempList, nums);
    tempList.pop();
  }
}
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
  const list = [];
  backtrack(list, [], nums);
  return list;
};

File: code/ch01/1.3.6.permute.java
class Solution {
  private void backtrack(
      List<List<Integer>> ans, ArrayList<Integer> tempList, ArrayList<Integer> nums) {
    if (tempList.size() == nums.size()) {
      ans.add(new ArrayList<Integer>(tempList));
      return;
    }
    for (int i = 0; i < nums.size(); i++) {
      if (tempList.contains(nums.get(i))) continue;
      tempList.add(nums.get(i));
      backtrack(ans, tempList, nums);
      tempList.remove(tempList.size() - 1);
    }
  }

  public List<List<Integer>> permute(int[] nums) {
    List<List<Integer>> ans = new LinkedList();
    ArrayList<Integer> tempList = new ArrayList<Integer>();
    ArrayList<Integer> anums = new ArrayList<Integer>();
    for (int num : nums) anums.add(num);
    backtrack(ans, tempList, anums);
    return ans;
  }
}
